mixer/adapter/circonus/circonus.go

```diff
@@ -0,0 +1,221 @@
+				Interval: "0s", // do not autoflush
+			}
+
+			var err error
+			cm, err = cgm.NewCirconusMetrics(cmc)
+			ec <- err
+		})
+
+	select {
+	case err := <-ec:
+		if err != nil {
+			env.Logger().Errorf("Error instantiating cgm: %v", err)
+			return nil, err
+		}
+	}
+
+	// wait until cgm background tasks initialized under ScheduleWork have completed
+	for !cm.Ready() {
+		time.Sleep(1 * time.Second)
+	}
+
+	metrics := make(map[string]config.Params_MetricInfo_Type)
+	ac := b.adpCfg
+	for _, adpMetric := range ac.Metrics {
+		metrics[adpMetric.Name] = adpMetric.Type
+	}
+
+	lastFlushed := time.Now()
+	flushInterval := b.adpCfg.SubmissionInterval
+
+	return &handler{cm: *cm, env: env, metrics: metrics, lastFlushed: lastFlushed, flushInterval: flushInterval}, nil
+}
+
+// SetAdapterConfig assigns operator configuration to the builder
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adpCfg = cfg.(*config.Params)
+}
+
+// Validate checks the SubmissionUrl parameter is well formed, and needed metrics are declared
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+
+	// verify the submission url is well formed
+	if _, err := url.ParseRequestURI(b.adpCfg.SubmissionUrl); err != nil {
+		ce = ce.Append("submission_url", err)
+	}
+
+	// put the metric config into a map for use below
+	configMetrics := make(map[string]config.Params_MetricInfo_Type)
+	for _, configMetric := range b.adpCfg.Metrics {
+		configMetrics[configMetric.Name] = configMetric.Type
+	}
+
+	// verify there are no metric types without a corresponding config item
+	for metricName := range b.metricTypes {
+		if _, ok := configMetrics[metricName]; !ok {
+			err := fmt.Errorf("missing metric configuration %v", metricName)
+			ce = ce.Append("metrics", err)
+		}
+	}
+
+	// verify there are no config items without a corresponding metric type
+	for configMetric := range configMetrics {
+		if _, ok := b.metricTypes[configMetric]; !ok {
+			err := fmt.Errorf("missing metric type for %v", configMetric)
+			ce = ce.Append("metrics", err)
+		}
+	}
+
+	// because SubmissionInterval is a duration type, validation errors will be caught before this point
+
+	return
+}
+
+// SetMetricTypes sets the available metric types in the builder
+func (b *builder) SetMetricTypes(types map[string]*metric.Type) {
+	b.metricTypes = types
+}
+
+// HandleMetric submits metrics to Circonus via circonus-gometrics
+func (h *handler) HandleMetric(ctx context.Context, insts []*metric.Instance) error {
+
+	for _, inst := range insts {
+
+		metricName := inst.Name
+		metricType := h.metrics[metricName]
+
+		switch metricType {
+
+		case config.GAUGE:
+			value, _ := inst.Value.(int64)
+			h.cm.Gauge(metricName, value)
+
+		case config.COUNTER:
+			h.cm.Increment(metricName)
+
+		case config.DISTRIBUTION:
+			value, _ := inst.Value.(time.Duration)
+			h.cm.Timing(metricName, float64(value))
+		}
+
+	}
+
+	// see if we should flush
+	h.flushLock.Lock()
```

Comment:
I think this should be in a goroutine that runs on a timer.

If

Reply:
So that was the approach I tried initially, but for some reason the metrics in the cm struct were absent when flush was called from the separate goroutine. I did quite a bit of digging but haven't yet been able to determine why that occurs. It only occurs when run under mixer; run as standalone, the state of cm metrics isn't cleared.

That's right that metrics will be buffered between HandleMetric calls, but under traffic within the flushInterval it shouldn't be an issue.

I can keep digging at the timer approach if you think that's best; maybe I can produce a test adapter that isolates the specific issue I ran into.